src/passes/OptimizeCasts.cpp

```diff
@@ -0,0 +1,219 @@
+//
+// An example of an important pattern this handles are itable calls:
+//
+//  (call_ref
+//    (ref.cast $actual.type
+//      (local.get $object)
+//    )
+//    (struct.get $vtable ..
+//      (ref.cast $vtable
+//        (struct.get $itable ..
+//          (local.get $object)
+//        )
+//      )
+//    )
+//  )
+//
+// We cast to the actual type for the |this| parameter, but we technically do
+// not need to do so for reading its itable - since the itable may be of a
+// generic type, and we cast the vtable afterwards anyhow. But since we cast
+// |this|, we can use the cast value for the itable get, which may then lead to
+// removing the vtable cast after we refine the itable type. And that can lead
+// to devirtualization later.
+//
+// Closely related things appear in other passes:
+//
+//  * SimplifyLocals will find locals already containing a more refined type and
+//    switch to them. RedundantSetElimination does the same across basic blocks.
+//    In theory one of them could be extended to also add new locals, and then
+//    they would be doing something similar to this pass.
+//  * LocalCSE finds repeated expressions and stores them in locals for use
+//    later. In theory that pass could be extended to look not for exact copies
+//    but for equivalent things through a cast, and then it would be doing
+//    something similar to this pass.
+//
+// However, while those other passes could be extended to cover what this pass
+// does, we will have further cast-specific optimizations to add, which make
+// sense in new pass anyhow, and things should be simpler overall to keep such
+// casts all in one pass, here.
+//
+// TODO: Move casts earlier in a basic block as well, at least in traps-never-
+//       happen mode where we can assume they don't happen.
+// TODO: Look past individual basic blocks?
+//
+
+#include "ir/linear-execution.h"
+#include "ir/properties.h"
+#include "ir/utils.h"
+#include "pass.h"
+#include "wasm-builder.h"
+#include "wasm.h"
+
+namespace wasm {
+
+namespace {
+
+// Find the best sources for local.gets: other locals with the same value, but
+// cast to a more refined type.
+struct BestSourceFinder : public LinearExecutionWalker<BestSourceFinder> {
+
+  PassOptions options;
+
+  // A map of the best local.get for a particular index: the local.get that has
+  // the most refined type.
+  std::unordered_map<Index, Expression*> bestSourceForIndexMap;
+
+  // For each expression, a vector of local.gets that would like to use its
+  // value instead of themselves (as it is more refined).
+  std::unordered_map<Expression*, std::vector<LocalGet*>> requestMap;
+
+  static void doNoteNonLinear(BestSourceFinder* self, Expression** currp) {
+    self->bestSourceForIndexMap.clear();
+  }
+
+  void visitLocalSet(LocalSet* curr) {
+    // Clear any information about this local; it has a new value here.
+    bestSourceForIndexMap.erase(curr->index);
+  }
+
+  void visitLocalGet(LocalGet* curr) {
```

Comment:
I would expect that we could also look at `LocalSet` here as long as we were careful about not looking through interfering tees. Maybe worth a TODO?